Implement data coordinator for Adax-integration

lazytarget opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This implements two DataCoordinators for the Adax-integration, one for each connection type (Cloud and Local).
This is something that has been requested as a dependency via another PR: #131314 

When fetching from Adax API, the results are stored in the Coordinator. 
Entities listen on changes in the coordinator via the `_handle_coordinator_update` callback function
If an entity requests updates then it will be flagged to the coordinator via the `await self.coordinator.async_request_refresh()`

`SCAN_INTERVAL` has been set to 60 seconds

Added a version to `manifest.json` since it was missing

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

lazytarget committed:
Implemented coordinator (for Cloud integration)

lazytarget committed:
Optimized coordinator updates

lazytarget committed:
Finalizing

lazytarget committed:
Running ruff and ruff format

lazytarget committed:
Raise error if trying to instantiate coordinator for a AdaxLocal config

Danielhiversen commented on homeassistant/components/adax/climate.py:
```diff
@@ -36,36 +37,30 @@ async def async_setup_entry(
+        async_add_entities(
+            [
+                LocalAdaxDevice(
+                    coordinator.adax_data_handler, entry.data[CONF_UNIQUE_ID]
```
Why do you use the coordinator for local mode?

Where is `adax_data_handler` defined for the local mode?

LazyTarget replied:
Oh, Nicely spotted!
Missed to pass a valid handler after doing a final refactoring

LazyTarget replied:
Fixed

lazytarget committed:
Merge branch 'dev' into feature/adax/data-coordinator

lazytarget committed:
Re-added data-handler for AdaxLocal integrations

Danielhiversen commented on homeassistant/components/adax/climate.py:
```diff
@@ -36,36 +39,31 @@ async def async_setup_entry(
-            entry.data[CONF_IP_ADDRESS],
-            entry.data[CONF_TOKEN],
-            websession=async_get_clientsession(hass, verify_ssl=False),
+        adax_data_handler = Adax(
```
```suggestion
        adax_data_handler = AdaxLocal(
```

LazyTarget replied:
Fixed in latest push

Danielhiversen commented on homeassistant/components/adax/__init__.py:
```diff
@@ -2,15 +2,23 @@
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Adax from a config entry."""
+    entry.coordinator = AdaxCoordinator(hass, entry, update_interval=SCAN_INTERVAL)
+    await entry.coordinator.async_config_entry_first_refresh()
```
This will also be called for the local integration?

LazyTarget replied:
Refactored and introduced a coordinator for the local mode as well.

Now the appropriate coordinator will be used based on config

lazytarget committed:
Added a coordinator for Local integrations

lazytarget committed:
mypy warnings

LazyTarget commented on homeassistant/components/adax/climate.py:
```diff
@@ -36,36 +39,31 @@ async def async_setup_entry(
+        adax_data_handler = Adax(
+            entry.data[ACCOUNT_ID],
+            entry.data[CONF_PASSWORD],
+            websession=async_get_clientsession(hass),
```
Fixed in latest push

LazyTarget commented:
**NOTE: I have only tested the Cloud mode. As I don't have the possiblity to test Local as of now**

Danielhiversen commented on homeassistant/components/adax/manifest.json:
```diff
@@ -6,5 +6,6 @@
   "iot_class": "local_polling",
   "loggers": ["adax", "adax_local"],
-  "requirements": ["adax==0.4.0", "Adax-local==0.1.5"]
+  "requirements": ["adax==0.4.0", "Adax-local==0.1.5"],
+  "version": "1.1.0"
```
```suggestion
  "requirements": ["adax==0.4.0", "Adax-local==0.1.5"]
```

Should only be used for custom components

lazytarget committed:
Update homeassistant/components/adax/manifest.json

Co-authored-by: Daniel Hjelseth Høyer <mail@dahoiv.net>

lazytarget committed:
Resolve mypy issues

lazytarget committed:
Merge remote-tracking branch 'lazytarget/feature/adax/data-coordinator' into feature/adax/data-coordinator

frenck commented on homeassistant/components/adax/__init__.py:
```diff
@@ -2,15 +2,31 @@
+    else:
+        coordinator = AdaxCloudCoordinator(hass, entry, update_interval=SCAN_INTERVAL)
+
+    hass.data[entry.entry_id] = coordinator
```
Please do not store the coordinator in hass.data, instead, use the config entry runtime_data, see:

https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/runtime-data

LazyTarget replied:
Applied as suggested! 👍

frenck commented on homeassistant/components/adax/coordinator.py:
```diff
@@ -0,0 +1,103 @@
+    ) -> None:
+        """Initialize the Adax coordinator used for Cloud mode."""
+        super().__init__(
+            hass, logger=_LOGGER, name="AdaxCloud", update_interval=update_interval
```
Make sure to explicitly pass the configuration entry on to the coordinator (this parameter is missing).

LazyTarget replied:
Updated in latest push

frenck commented on homeassistant/components/adax/coordinator.py:
```diff
@@ -0,0 +1,103 @@
+        """Fetch data from the Adax."""
+        try:
+            self.rooms = await self.adax_data_handler.get_rooms()
+        except Exception as err:
```
We shouldn't be catching bare exceptions, please make this more specific.

LazyTarget replied:
Removed try-catch entirely. I guess the exception type should not matter in this case, then the only other difference is that that log statement will be removed (but the uncaught exception will still be propagated up and logged from there)

frenck commented on homeassistant/components/adax/coordinator.py:
```diff
@@ -0,0 +1,103 @@
+        try:
+            self.rooms = await self.adax_data_handler.get_rooms()
+        except Exception as err:
+            _LOGGER.error("Exception when getting data. Err: %s", err)
```
This is already logged by raising the exception below? I think this can be removed.

LazyTarget replied:
Removed

frenck commented on homeassistant/components/adax/coordinator.py:
```diff
@@ -0,0 +1,103 @@
+    async def _async_update_data(self) -> list[dict[str, Any]]:
+        """Fetch data from the Adax."""
+        try:
+            self.rooms = await self.adax_data_handler.get_rooms()
```
We store in `self.room` but also return it in the end (thus ending up in self.data).

Why are we storing this data twice?

LazyTarget replied:
I did not know this. Very good, updated in latest push

frenck commented on homeassistant/components/adax/coordinator.py:
```diff
@@ -0,0 +1,103 @@
+    ) -> None:
+        """Initialize the Adax coordinator used for Local mode."""
+        super().__init__(
+            hass, logger=_LOGGER, name="AdaxLocal", update_interval=update_interval
```
Same as above, we should pass the config entry explicitly.

LazyTarget replied:
Updated here as well

frenck commented on homeassistant/components/adax/coordinator.py:
```diff
@@ -0,0 +1,103 @@
+        """Fetch data from the Adax."""
+        try:
+            self.status = await self.adax_data_handler.get_status()
+        except Exception as err:
```
Do not catch bare exceptions, make it more specific.

LazyTarget replied:
Try-catch removed

frenck commented on homeassistant/components/adax/coordinator.py:
```diff
@@ -0,0 +1,103 @@
+        try:
+            self.status = await self.adax_data_handler.get_status()
+        except Exception as err:
+            _LOGGER.error("Exception when getting data. Err: %s", err)
```
Same as above, shouldn't be needed.

LazyTarget replied:
Try-catch removed

frenck commented on homeassistant/components/adax/coordinator.py:
```diff
@@ -0,0 +1,103 @@
+    async def _async_update_data(self) -> dict[str, Any]:
+        """Fetch data from the Adax."""
+        try:
+            self.status = await self.adax_data_handler.get_status()
```
Same as above, we store this data twice now.

LazyTarget replied:
Updated as per suggestions above

lazytarget committed:
PR comments

- Explicit passing of config_entry to Coordinator base type
- Avoid duplicate storing of Coordinator data. Instead use self.data
- Remove try-catch and wrapping to UpdateFailed in _async_update_data
- Custom ConfigEntry type for passing coordinator via entry.runtime_data

lazytarget committed:
When changing HVAC_MODE update data via Coordinator to optimize

lazytarget committed:
Apply already loaded data for Climate entity directly in __init__

frenck commented on homeassistant/components/adax/__init__.py:
```diff
@@ -2,16 +2,38 @@
+async def async_setup_entry(hass: HomeAssistant, entry: AdaxConfigEntry) -> bool:
     """Set up Adax from a config entry."""
+    if entry.data.get(CONNECTION_TYPE) == LOCAL:
+        coordinator = AdaxLocalCoordinator(hass, entry, update_interval=SCAN_INTERVAL)
+    else:
+        coordinator = AdaxCloudCoordinator(hass, entry, update_interval=SCAN_INTERVAL)
```
If the update interval is the same for both, we can also just not pass it in, and just set it directly on both coordinators. (Move SCAN_INTERVAL into the const module).

LazyTarget replied:
Pushed update

frenck commented on homeassistant/components/adax/__init__.py:
```diff
@@ -2,16 +2,38 @@
+        coordinator = AdaxCloudCoordinator(hass, entry, update_interval=SCAN_INTERVAL)
+
+    entry.runtime_data = coordinator
+    _LOGGER.info("Init Adax, coordinator: %s", coordinator)
```
This is not something we have to log (especially not @ info level)
```suggestion
```

LazyTarget replied:
All testing log statements have now been removed 😄

frenck commented on homeassistant/components/adax/__init__.py:
```diff
@@ -2,16 +2,38 @@
+    entry.runtime_data = coordinator
+    _LOGGER.info("Init Adax, coordinator: %s", coordinator)
+    await coordinator.async_config_entry_first_refresh()
+    _LOGGER.info("First refresh complete: %s", coordinator.get_rooms())
```
Same, if we need to get data from the coordinator, we should expose that via the diagnostic panel.

```suggestion
```

frenck commented on homeassistant/components/adax/__init__.py:
```diff
@@ -2,16 +2,38 @@
+    _LOGGER.info("First refresh complete: %s", coordinator.get_rooms())
+
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    _LOGGER.info("Setup complete")
```
```suggestion
```

frenck commented on homeassistant/components/adax/climate.py:
```diff
@@ -12,60 +13,52 @@
+    else:
+        coordinator: AdaxCloudCoordinator = entry.runtime_data
+        rooms = coordinator.get_rooms()
+        _LOGGER.info("Setup entries: %s", rooms)
```
This is already logged... 

```suggestion
```

frenck commented on homeassistant/components/adax/coordinator.py:
```diff
@@ -0,0 +1,95 @@
+            websession=async_get_clientsession(hass, verify_ssl=False),
+        )
+
+    def get_status(self) -> dict[str, Any]:
+        """Get status for the Adax device."""
+        return self.data
+
```
Why do we need `self.get_status()` to get `self.data`? We can just use self.data directly everywhere instead, right?

LazyTarget replied:
Replaced those instances to `self.data`

lazytarget committed:
Moved SCAN_INTERVAL into const.py

lazytarget committed:
Removed logging statements

lazytarget committed:
Remove unnecessary get_rooms() / get_status() functions

frenck reviewed (dismissed):
The CI is failing right now, could you please take a look?

Thanks! 👍 

../Frenck

lazytarget committed:
Resolvning mypy issues

LazyTarget commented:
I belive these changes also solves this issue:
#131007 

The issue there is that the warning"Max 1 request per 30 second" is logged.
This can be reproduced in the `dev`-branch by enabling debug logging for `adax` to see when the requests are being made and when updates are skipped:
![image](https://github.com/user-attachments/assets/6a8dd7b4-a660-4546-aae2-6c5772de554f)

Then change the temperature on one of your devices within the 30 second mark.
![image](https://github.com/user-attachments/assets/195c8f95-afdf-4cdf-9b12-c45cfb3a5274)

When I try to reproduce this with the PR changes the warning-message never gets logged

lazytarget committed:
Adding tests for coordinators

lazytarget committed:
Resolving review comments by joostlek

lazytarget committed:
Setup of Cloud devices with device_id

lazytarget committed:
Implement Climate tests for Adax

lazytarget committed:
Implementing assertions of UNAVAILABLE state

LazyTarget commented on tests/components/adax/test_coordinator.py:
```diff
@@ -0,0 +1,98 @@
+        mock_adax.get_rooms = AsyncMock(side_effect=Exception)
+        with pytest.raises(UpdateFailed):
+            await coordinator.async_refresh()
+
```
Updated tests! 😄 👍

LazyTarget commented:
@joostlek 
Updated according to your suggestions
Also updated the tests and made sure they all pass locally
Tested Cloud integration locally and data is loaded successfully

lazytarget committed:
Removed no longer needed method

joostlek commented on homeassistant/components/adax/climate.py:
```diff
@@ -38,22 +38,21 @@ async def async_setup_entry(
         async_add_entities(
-            (AdaxDevice(room, cloud_coordinator) for room in cloud_coordinator.data),
-            True,
+            (
+                AdaxDevice(cloud_coordinator, device_id)
+                for device_id in cloud_coordinator.data
+            ),
```
```suggestion
            AdaxDevice(cloud_coordinator, device_id)
            for device_id in cloud_coordinator.data
```

joostlek commented on homeassistant/components/adax/climate.py:
```diff
@@ -65,23 +64,36 @@ class AdaxDevice(CoordinatorEntity[AdaxCloudCoordinator], ClimateEntity):
+        self._attr_unique_id = f"{self.room['homeId']}_{self._device_id}"
         self._attr_device_info = DeviceInfo(
-            identifiers={(DOMAIN, heater_data["id"])},
+            identifiers={(DOMAIN, self._device_id)},
```
```suggestion
            identifiers={(DOMAIN, device_id)},
```

joostlek commented on tests/components/adax/__init__.py:
```diff
@@ -1 +1,44 @@
+}
+
+
+async def init_integration(
+    hass: HomeAssistant, entry_data: dict[str, Any]
+) -> MockConfigEntry:
+    """Set up the Adax integration in Home Assistant."""
+    entry = MockConfigEntry(domain=DOMAIN, data=entry_data)
+    entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(entry.entry_id)
+    await hass.async_block_till_done()
+    return entry
```
Instead of having the entry_data as a parameter, why don't we create 2 MockConfigEntries as fixtures? Check `airgradient` for an example

joostlek commented on homeassistant/components/adax/coordinator.py:
```diff
@@ -0,0 +1,71 @@
+            websession=async_get_clientsession(hass, verify_ssl=False),
+        )
+
+    async def _async_update_data(self) -> dict[str, Any] | None:
```
```suggestion
    async def _async_update_data(self) -> dict[str, Any]:
```

joostlek commented on tests/components/adax/conftest.py:
```diff
@@ -0,0 +1,38 @@
+@pytest.fixture
+def mock_adax_cloud():
+    """Mock climate data."""
+    with patch("adax.Adax.get_rooms") as mock_adax_get_rooms:
+        mock_adax_get_rooms.return_value = CLOUD_DEVICE_DATA
+        yield mock_adax_get_rooms
+
+
+@pytest.fixture
+def mock_adax_local():
+    """Mock climate data."""
+    with patch("adax_local.Adax.get_status") as mock_adax_class_get_status:
```
we should also patch this at the place where we use it. I would recommend to patch out the object rather than the method

LazyTarget replied:
For some reason I was not able to get it to work when I patched the object. Probably something I am doing wrong or wront with my local setup.

Feels like the Adax library is imported and cached before the patch. When resolving the type, only the concrete type is resolved

joostlek replied:
Where did you patch it? You have to patch it to where it was initially created. I'm also not sure if adax imports with `import X as xy`

LazyTarget replied:
I was able to patch the Adax library as it is loaded via `homeassistant.components.adax.coordinator.`
Tests have been updated and seem to correctly patch the relevant members.

PvaI commented:
Hello gents, I've pulled this code and tested with 5 Adax Heaters (Neo & Clea) on 2025.4.2. 

Observations: 
 - Changing a setting shows up in the mobile app within 30 seconds.
 - HA can be just as quick, but at other times is quite a bit slower.
 - HA changes back to the previous setting after a minute or so and later changes back to the desired setting after 2 - 4 minutes.
 - The device only seems to poll the Adax servers every 10 minutes

e.g. times in seconds
```Change target temperature in HA:
  0 - Change setting in HA
 30 - Mobile app shows new target temp 
 46 - HA shows new target temp 
 78 - HA shows previous target temp
132 - HA shows new target temp again

Turn heater on in HA:
  0 - Change setting in HA
 16 - HA shows heater turned on
 20 - Mobile app shows heater turned on
 66 - HA shows heater turned off
187 - HA shows heater turned on

Turn off heater in HA:
  0 - Change setting in HA
  9 - Setting shows up in mobile app and HA almost simultaneously
 57 - HA shows heater as being turned on
266 - HA shows heater as being turned off
```

We seem to be using a rate limit and a scan interval, could they be interfering?
pyAdax : RATE_LIMIT_SECONDS = 30
const.py : SCAN_INTERVAL = datetime.timedelta(seconds=60)

Adax code example and OpenHAB implementation seem to use a 10 second delay:
Adax API code sample: https://adax.no/en/wi-fi/api-development/
OpenHAB: https://community.openhab.org/t/adax-panel-heater-with-wifi-thermostat-connected-to-openhab-using-the-http-binding

OpenHAB dev mentions: 
 - the refresh interval is a factor that can provoke some instability, don’t set it to short.
 - Little easteregg on 29/03/2028; the Adax dev people managed to sneak a leaptimebug in to the local api stack, unless they manage to remove the bug by a firmware upgrade, you’ll have to reregister/integrate the heaters, it’s a bit anoying but it’s a once in a four year chore
 
fyi: local API on newer models: https://adax.no/wi-fi/api-development/#local (only visible when you set the language to Norwegian)

lazytarget committed:
Apply suggestions from code review

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

lazytarget committed:
Mock Adax class instead of individual methods

lazytarget committed:
Mock config entry via fixture

lazytarget committed:
Load config entry data from .json fixture

joostlek commented on tests/components/adax/conftest.py:
```diff
@@ -22,17 +28,34 @@
+@pytest.fixture(params=["cloud", "local"])
+def mock_config_entry(request: pytest.FixtureRequest) -> MockConfigEntry:
+    """Mock a config entry."""
+    f = load_fixture(f"get_config_{request.param}.json", DOMAIN)
+    entry_data = json.loads(f)
```
Oh right, with fixture I mean, we can create a `mock_local_config_entry` and `mock_cloud_config_entry` and then use that where we want

LazyTarget replied:
Reverted the json-file fixtures, to instead use hardcoded JSON objects

lazytarget committed:
Hard code config_entry_data instead of .json file

lazytarget committed:
Removed obsolete .json-files

LazyTarget commented:
> Hello gents, I've pulled this code and tested with 5 Adax Heaters (Neo & Clea) on 2025.4.2.
> 
> Observations:
> 
> * Changing a setting shows up in the mobile app within 30 seconds.
> * HA can be just as quick, but at other times is quite a bit slower.
> * HA changes back to the previous setting after a minute or so and later changes back to the desired setting after 2 - 4 minutes.
> * The device only seems to poll the Adax servers every 10 minutes

Hello @PvaI,
Thank you for your observations! I came to pretty much the same conclusions while getting my hands dirty with the Adax-api's.
I have a few Neo Wifi 1000W and 600W myself. Running on firmware 2.0.2.29 (and one that is plugged out, on 2.0.2.26)


> We seem to be using a rate limit and a scan interval, could they be interfering? pyAdax : RATE_LIMIT_SECONDS = 30 const.py : SCAN_INTERVAL = datetime.timedelta(seconds=60)

I don't believe that the combination of Rate limit and Scan interval is the reason, to why temperatures switches back and forth. My feeling is that the Adax app tells the cloud heater to set a target temperature, the value is persisted in the cloud and notifies the device directly or via a shorter polling. But for HA integrations (`pyAdax`) the target temperature is also set in the cloud, but on a delay, possibly queued for persistance? Or that the direct communication to the device is not established so that the local device doesn't yet know about the temperature change. Will try to debug more and wrap my head around it, was a few months ago now that I investigated this.


> * Little easteregg on 29/03/2028; the Adax dev people managed to sneak a leaptimebug in to the local api stack, unless they manage to remove the bug by a firmware upgrade, you’ll have to reregister/integrate the heaters, it’s a bit anoying but it’s a once in a four year chore

Do you know if they have fixed this in a fireware upgrade by now?



> fyi: local API on newer models: https://adax.no/wi-fi/api-development/#local (only visible when you set the language to Norwegian)

It's weird... In a way they have good documentation on the API. On the other way it is horrible! Their code examples don't work out of the box either

joostlek committed:
Fix

joostlek committed:
Merge branch 'dev' into feature/adax/data-coordinator

joostlek committed:
Fix
